server/testutils/factories/courseParticipation.ts

```diff
@@ -3,7 +3,6 @@ import { Factory } from 'fishery'
 
```

Comment:
As discussed, might be good to clarify "nullable or string data" as `string | undefined` (as opposed to `string`)

Reply:
Makes sense - have updating the commit message to include this and the Pact info in your other comment - hope it's a bit clearer!

Reply:
Is it: for optional string fields, Pact either wants a string or nothing, not `null` or `undefined`? The problem being we're sometimes sending `null` or `undefined` when it wants a string or nothing?

Reply:
What happens is:

- UI client test generates an object with an undefined value (e.g. `CourseParticipation.setting`), then generates the JSON contract file expecting the API to send an object with the undefined value (or the field missing), e.g - second object here is missing the `setting` object:

```
{
  "status": 200,
  "headers": {
    "Content-Type": "application/json"
  },
  "body": [
    {
      "addedBy": "Devin Bins",
      "courseName": "Red Course",
      "createdAt": "Fri Nov 24 2023 03:46:19 GMT+0000 (Coordinated Universal Time)",
      "detail": "",
      "id": "0cff5da9-1e90-4ee2-a5cb-94dc49c4b004",
      "outcome": {
        "status": "incomplete",
        "yearStarted": 1995
      },
      "prisonNumber": "A1234AA",
      "setting": {
        "location": "Long Funk",
        "type": "custody"
      },
      "source": "rightfully"
    },
    {
      "addedBy": "Trace Nikolaus",
      "courseName": "Yellow Course",
      "createdAt": "Tue Nov 07 2023 05:09:08 GMT+0000 (Coordinated Universal Time)",
      "detail": "Confido charisma clamo consectetur.",
      "id": "eb357e5d-5416-43bf-a8d2-0dc8fd92162e",
      "outcome": {
        "status": "incomplete"
      },
      "prisonNumber": "A1234AA",
      "source": "usefully outside where"
    }
  ]
}
```

- The API fetches the generated JSON contract and plays the values in the [seed data sql file](https://github.com/ministryofjustice/hmpps-accredited-programmes-api/blob/550d2ec6d814272977fb4d02bddf21549a097bc6/src/test/resources/db/migration/R__test_data.sql#L13) against the contract generated above. Because none of those entities have a `null` `location` (what we call `setting` on the backend), the API tests fail, as the UI is apparently expecting the entity not to have the `setting` field on it.

Reply:
Hmm. Would generating data in our UI client tests that match the API test data not fix this problem then, or extending the test data where needed? I know we want to move away from the SQL file at some point anyway though

Reply:
It's interesting that it omits the `setting` rather than setting it as `null` though